File: algorithms/sorting/cocktail-sort/python/cocktail_sort.py
def cocktail_sort(arr: list[int]) -> list[int]:
    """
    Cocktail Sort (Bidirectional Bubble Sort) implementation.
    Repeatedly steps through the list in both directions, comparing adjacent elements 
    and swapping them if they are in the wrong order.
    """
    result = list(arr)
    n = len(result)
    if n <= 1:
        return result

    start = 0
    end = n - 1
    swapped = True

    while swapped:
        swapped = False

        # Forward pass (like bubble sort)
        for i in range(start, end):
            if result[i] > result[i + 1]:
                result[i], result[i + 1] = result[i + 1], result[i]
                swapped = True

        if not swapped:
            break

        swapped = False
        # Last element is now in place
        end -= 1

        # Backward pass
        for i in range(end - 1, start - 1, -1):
            if result[i] > result[i + 1]:
                result[i], result[i + 1] = result[i + 1], result[i]
                swapped = True

        # First element is now in place
        start += 1

    return result

File: algorithms/sorting/cocktail-sort/go/cocktail_sort.go
package cocktailsort

/**
 * CocktailSort implementation.
 * Repeatedly steps through the list in both directions, comparing adjacent elements 
 * and swapping them if they are in the wrong order.
 * It returns a new sorted slice without modifying the original input.
 */
func CocktailSort(arr []int) []int {
	n := len(arr)
	if n <= 1 {
		return append([]int{}, arr...)
	}

	result := make([]int, n)
	copy(result, arr)

	start := 0
	end := n - 1
	swapped := true

	for swapped {
		swapped = false

		// Forward pass
		for i := start; i < end; i++ {
			if result[i] > result[i+1] {
				result[i], result[i+1] = result[i+1], result[i]
				swapped = true
			}
		}

		if !swapped {
			break
		}

		swapped = false
		end--

		// Backward pass
		for i := end - 1; i >= start; i-- {
			if result[i] > result[i+1] {
				result[i], result[i+1] = result[i+1], result[i]
				swapped = true
			}
		}

		start++
	}

	return result
}
